### PR TITLE
[OP] Enable register via match tag

### PR DESCRIPTION
**Pull request by tqchen**
(edit) new api

```c++

NNVM_REGISTER_OP(add)
.include("ElementwiseOpAttr");

// register same attributes to all the ops that matchs the attribute condition.
NNVM_REGISTER_OP_GROUP(ElementwiseOpAttr)
.set_attr<FInferShape>("FInferShape", ElementwiseInferShape);

// mul will also have the ElementwiseInferShape 
NNVM_REGISTER_OP(mul)
.include("ElementwiseOpAttr");
```

### CODE REVIEW

**Comment by tqchen** [comment]
@jermainewang @piiswrong

**Comment by tqchen** [comment]
please see if the name  ```NNVM_REGISTER_BY_MATCH_ATTR``` makes sense to you, I think it is an OK name and cannot come up with a shorter version for now

**Comment by piiswrong** [comment]
```
NNVM_REGISTER_OP(add)
.set_attr<std::string>("OpType", "Elementwise");

// register same attributes to all the ops that matchs the attribute condition.
NNVM_REGISTER_OP_GROUP(OpType, "Elementwise")
.set_attr<FInferShape>("FInferShape", ElementwiseInferShape);

// mul will also have the ElementwiseInferShape tag
NNVM_REGISTER_OP(mul)
.set_attr<std::string>("OpGroup", "Elementwise");
```
looks better

**Comment by piiswrong** [comment]
BTW, what's the relation between OpType and TagType? It's confusing

**Comment by piiswrong** [comment]
This makes it even harder to figure out where things are registered.

It would be MUCH easier to document and understand if you use a different interface for setting tags:
```
NNVM_REGISTER_OP(mul)
.add_to_group("Elementwise");
```
also instead of using names like Elementwise, it should be something unique so that a search in the code base will point to the registration point

**Comment by piiswrong** [comment]
Actually I think
```
NNVM_REGISTER_OP(mul)
.include("ElemwiseOpAttrs");
```
Makes more sense

**Comment by tqchen** [comment]
The idea is that we don't want to specialize the field group or tag. But simply use attribute. There can be multiple attributes that corresponding to a tag.

For example
```
NNVM_REGISTER_OP(add)
.set_attr<std::string>("OpType", "Elementwise")
.set_attr<std::string>("FuncType", "BinaryOp");

// register same attributes to all the ops that matchs the attribute condition.
NNVM_OP_GROUP_BY_MATCH_ATTR(OpType, "Elementwise")
.set_attr<FInferShape>("FInferShape", ElementwiseInferShape);

NNVM_OP_GROUP_BY_MATCH_ATTR(FuncType, "UnaryOp")
.set_attr<FXX>("FXX", UnaryFXX);

// mul will also have the ElementwiseInferShape tag
NNVM_REGISTER_OP(mul)
.set_attr<std::string>("OpType", "Elementwise");

NNVM_REGISTER_OP(exp)
.set_attr<std::string>("OpType", "Elementwise")
.set_attr<std::string>("FuncType", "Unary");


```

**Comment by piiswrong** [comment]
That makes it more complicated but doesn't add much utility. `include(OpType:Elementwise)` would to the same thing.

I think it definitely need a different api. you can still implement it via attrs internally.

**Comment by piiswrong** [comment]
Consider doing it this way: an op can include another op and copy all it's attrs. This can be done recursively and groups can also be registered as ops

**Comment by tqchen** [comment]
Group is definitely different from what op is(they have incomplete information, and partial info about something), so it does not make sense to do so.

```NNVM_OP_GROUP_BY_MATCH_ATTR``` provides a semantics like

```c++
NNVM_FOR_EACH_OP_WHOSE_ATTR_EQ(AttrName, Attr_Value)
DO
op.set_attr<bla>()
END
```
Which is quite clear,

**Comment by piiswrong** [comment]
consider ops as classes. an incomplete abstract class is still a class

**Comment by tqchen** [comment]
There are two semantics, here, both are valid, the difference is like follows

- Inheritate provides a ***pull semantics***, operators pull things from its parents,
    - There could be confusion about multiple inheritance in here
    - The recursive inheritance seems to be a overkill feature that is a bit hard to reason
    - It is not easy to tell from an abstract class from normal one, especially when multiple inheritance happen. 
- Match and set provides a ***push semantics***, which means find the ops that satisfies the condition and set the corresponding attribute.

Personally I prefer not introducing group as op, but simply as a returning result of some matching conditions on the operator set, which is aligned with push semantics. It also avoid introducing additional concept for specially handle groups.

**Comment by tqchen** [comment]
@piiswrong changed API to include

File: src/core/op.cc [hunk 1]
```diff
@@ -66,10 +72,42 @@ const any* Op::GetAttrMap(const std::string& key) {
...
+  auto& op_group = mgr->op_group;
+  for (const Op* op : dmlc::Registry<Op>::List()) {
+    if (op->index_ < op_group.size() &&
+        op_group[op->index_].count(group_name) != 0) {
+      trigger((Op*)op);  // NOLINT(*)
+    }
+  }
+}
+
+Op& Op::include(const std::string& group_name) {
```
Comment (piiswrong) [inline src/core/op.cc]: document this?

Comment (tqchen) [inline src/core/op.cc]: documents in the header

Comment (tqchen) [inline src/core/op.cc]: https://github.com/dmlc/nnvm/pull/57/files#diff-c73a5e795727ea0120db27bc6d1b9bd1R234

Comment (tqchen) [inline src/core/op.cc]: https://github.com/tqchen/nnvm/blob/master/include/nnvm/op.h#L234

**Pull request merged**